common/src/autogluon/common/__init__.py

```diff
@@ -1,6 +1,7 @@
 from .version import __version__
 
 from .utils.log_utils import _add_stream_handler, fix_logging_if_kaggle as __fix_logging_if_kaggle
+from .space import Space, Categorical, Real, Int, Bool
```

Comment:
Would it be possible to avoid a breaking change by keeping the spaces available in `autogluon.core` for 1 more release with a deprecation warning?

For example, using something like the following code in `autogluon/core/__init__.py`
```python
import warnings

class DeprecatedSpacesWrapper:
    def __getattr__(self, attr: str) -> Any:
        import autogluon.common as ag

        if attr in ["Categorical", "Real", "Int", "Bool"]:
            warnings.warn(
                "Accessing search spaces as `autogluon.core.space` is deprecated as of v0.8 and won't be supported "
                "in the next release. Please use `autogluon.common.space` instead."
            )
            return getattr(ag.space, attr)
        else:
            raise AttributeError

space = DeprecatedSpacesWrapper()
```

Reply:
Thanks for bringing it up. I think this is a good idea

Reply:
+1, this is nice

Reply:
I moved the warning to the init of any `Space` object and imported `Space` related object from common in core instead.

The method @shchur provided wouldn't work if user try to do things like `from autogluon.core import Real`, which is possible currently. After the change, the above usage would also give proper warning message

Reply:
Would this result in a warning always being produced when a `Space` is used (e.g., as a part of the presets), even if the user didn't use the deprecated API?

Reply:
Oh yea... This is tricky. Need to think about how to cover all cases

Reply:
OK. Added this inside `core.space.py`, and this plus the wrapper @shchur provided should cover all cases and not affect import from common

```python
# TODO: Remove this file for v1.0 as space is moved to common
import warnings
import sys

from typing import Any


this_module = sys.modules[__name__]
spaces = ["Space", "Categorical", "Real", "Int", "Bool"]


class DeprecatedSpaceWrapper:
    def __init__(self, common_space) -> None:
        self.comomn_space = common_space
    
    def __call__(self, *args, **kwargs) -> Any:
        import autogluon.common as ag

        warnings.warn(
            "Accessing search spaces through `autogluon.core` is deprecated as of v0.8 and won't be supported "
            f"in the next release. Please use `from autogluon.common import {self.comomn_space}` instead."
        )
        return getattr(ag.space, self.comomn_space)(*args, **kwargs)


for s in spaces:
    setattr(this_module, s, DeprecatedSpaceWrapper(s))

```